### PR TITLE
New: Add minimum version checks for applications

### LINKED ISSUE #1674: Add Version Checks for Radarr/Sonarr
### Is there an existing issue for this?

- [X] I have searched the existing open and closed issues

### Is your feature request related to a problem? Please describe

Users are running old unsupported versions of Radarr and Sonarr - check and verify the version endpoint (for v3) and if the v3 api does not exist (for v2 and v0.2) and error accordingly

only needs to occur on application test / first add.

### Describe the solution you'd like

see above

### Describe alternatives you've considered

weird error spew for unsupported versions about reference issues which is unclear and confusing to the user

### Anything else?

n/a - some of this already exists

### PR DESCRIPTION
**Pull request by mynameisbogdan**
#### Database Migration
NO

#### Issues Fixed or Closed by this PR

* Fixes #1674

### CODE REVIEW

File: src/NzbDrone.Core/Applications/Radarr/RadarrV3Proxy.cs [hunk 1]
```diff
@@ -102,7 +102,17 @@ public ValidationFailure TestConnection(RadarrIndexer indexer, RadarrSettings se
...
             {
-                Execute<RadarrIndexer>(request);
+                var version = _httpClient.Post<RadarrIndexer>(request).Headers.GetSingleValue("X-Application-Version");
+
+                if (version == null)
+                {
+                    return new ValidationFailure(string.Empty, "Failed to fetch Radarr version");
+                }
+
+                if (new Version(version) < new Version("4.0.0.0"))
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Applications/Radarr/RadarrV3Proxy.cs:112]: first v4 stable is `[4.0.4.5922](https://github.com/Radarr/Radarr/releases/tag/v4.0.4.5922)`

File: src/NzbDrone.Core/Applications/Lidarr/LidarrV1Proxy.cs [hunk 2]
```diff
@@ -102,7 +102,17 @@ public ValidationFailure TestConnection(LidarrIndexer indexer, LidarrSettings se
...
             {
-                Execute<LidarrIndexer>(request);
+                var version = _httpClient.Post<LidarrIndexer>(request).Headers.GetSingleValue("X-Application-Version");
+
+                if (version == null)
+                {
+                    return new ValidationFailure(string.Empty, "Failed to fetch Lidarr version");
+                }
+
+                if (new Version(version) < new Version("1.0.0.0"))
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Applications/Lidarr/LidarrV1Proxy.cs:112]: first lidarr v1 is [1.0.2.2592](https://github.com/Lidarr/Lidarr/releases/tag/v1.0.2.2592)

else current is [1.1.4.3027](https://github.com/Lidarr/Lidarr/releases/tag/v1.1.4.3027)

File: src/NzbDrone.Core/Applications/Sonarr/SonarrV3Proxy.cs [hunk 3]
```diff
@@ -102,7 +102,17 @@ public ValidationFailure TestConnection(SonarrIndexer indexer, SonarrSettings se
...
             {
-                Execute<SonarrIndexer>(request);
+                var version = _httpClient.Post<SonarrIndexer>(request).Headers.GetSingleValue("X-Application-Version");
+
+                if (version == null)
+                {
+                    return new ValidationFailure(string.Empty, "Failed to fetch Sonarr version");
+                }
+
+                if (new Version(version) < new Version("3.0.9.0"))
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Applications/Sonarr/SonarrV3Proxy.cs:112]:
```suggestion
                if (new Version(version) < new Version("3.0.9.1549"))
```

Comment (bakerboy448) [reply src/NzbDrone.Core/Applications/Sonarr/SonarrV3Proxy.cs:112]: let's move the version to a var to make it easier to maintain

Comment (bakerboy448) [reply src/NzbDrone.Core/Applications/Sonarr/SonarrV3Proxy.cs:112]: builds have never - in the ~12 years of starr have never been reset within a major version

File: src/NzbDrone.Core/Applications/Lidarr/LidarrV1Proxy.cs [hunk 4]
```diff
@@ -102,7 +104,17 @@ public ValidationFailure TestConnection(LidarrIndexer indexer, LidarrSettings se
...
+                var version = _httpClient.Post<LidarrIndexer>(request).Headers.GetSingleValue("X-Application-Version");
+
+                if (version == null)
+                {
+                    return new ValidationFailure(string.Empty, "Failed to fetch Lidarr version");
+                }
+
+                if (new Version(version) < new Version(MinimumApplicationVersion))
+                {
+                    return new ValidationFailure(string.Empty, $"Lidarr version should be at least 1.0.2. Version reported is {version}", version);
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Applications/Lidarr/LidarrV1Proxy.cs:116]: MinimumApplicationVersion in validation error :)

File: src/NzbDrone.Core/Applications/Sonarr/SonarrV3Proxy.cs [hunk 5]
```diff
@@ -102,7 +104,17 @@ public ValidationFailure TestConnection(SonarrIndexer indexer, SonarrSettings se
 
             try
             {
-                Execute<SonarrIndexer>(request);
```
Comment (bakerboy448) [inline src/NzbDrone.Core/Applications/Sonarr/SonarrV3Proxy.cs:105]: we still need this execution attempt to ensure the apps can talk; no?

File: src/NzbDrone.Core/Applications/Lidarr/LidarrV1Proxy.cs (hunk 4 above)
Comment (bakerboy448) [reply src/NzbDrone.Core/Applications/Lidarr/LidarrV1Proxy.cs:116]: `$"Lidarr version should be at least {MinimumApplicationVersion}. Version reported is {version}")` no?

**Comment by Qstick** [comment]
I'll have a look, I don't think there were breaking changes with provider adding going back to Sonarr 3 or Radarr 3. Readarr would likely be 0.1 and up, Lidarr 0.7+

**Comment by bakerboy448** [comment]
v3.0.4 was a beta sonarr version, so start with the first stable v3 and work forwards?

**Comment by bakerboy448** [comment]
IMO givens sonarr 3.0.4 is a legacy, unsupported beta. It shouldn't be supported here.

**Comment by bakerboy448** [comment]
[v3.0.5.1144](https://github.com/Sonarr/Sonarr/releases/tag/v3.0.5.1144)

**Comment by Qstick** [comment]
Lidarr 0.8 is still 10% of installs. That's the only other one I'd look at.

**Review by bakerboy448 — DISMISSED** [review]

File: src/NzbDrone.Core/Applications/Radarr/Radarr.cs [hunk 6]
```diff
@@ -174,7 +173,7 @@ private RadarrIndexer BuildRadarrIndexer(IndexerDefinition indexer, DownloadProt
         {
             var cacheKey = $"{Settings.BaseUrl}";
             var schemas = _schemaCache.Get(cacheKey, () => _radarrV3Proxy.GetIndexerSchema(Settings), TimeSpan.FromDays(7));
-            var syncFields = new[] { "baseUrl", "apiPath", "apiKey", "categories", "minimumSeeders", "seedCriteria.seedRatio", "seedCriteria.seedTime" };
+            var syncFields = new[] { "baseUrl", "apiPath", "apiKey", "categories", "multiLanguages", "removeYear", "requiredFlags", "additionalParameters", "minimumSeeders", "seedCriteria.seedRatio", "seedCriteria.seedTime" };
```
Comment (Qstick) [inline src/NzbDrone.Core/Applications/Radarr/Radarr.cs:176]: We shouldn’t add things to these that Prowlarr doesn’t control. I believe it’ll cause Prowlarr to override user settings set in arr

Comment (bakerboy448) [reply src/NzbDrone.Core/Applications/Radarr/Radarr.cs:176]: knew there was a reason those were removed.
would it be better if they were explicitly listed e.g. `noSyncFields` (with default noSync if undefined field in schema)?
perhaps that can then translate to some fancy UI info text to the users for what fields Full Sync evaluates

Comment (bakerboy448) [reply src/NzbDrone.Core/Applications/Radarr/Radarr.cs:176]: I fully support dropping older versions lol.

radarr v3 even is mentioned continually on discord as unsupported by staff

**Review by bakerboy448 — COMMENTED** [review]
Looks sane

**Review by Qstick — APPROVED** [review]

**Pull request merged**